text/0000-experimental-coroutines.md

```diff
@@ -0,0 +1,471 @@
+- Feature Name: `coroutines`
+- Start Date: 2017-06-15
+- RFC PR: (leave this empty)
+- Rust Issue: (leave this empty)
+
+# Summary
+[summary]: #summary
+
+This is an **experimental RFC** for adding a new feature to the language,
+coroutines (also commonly referred to as generators). This RFC is intended to be
+relatively lightweight and bikeshed free as it will be followed by a separate
+RFC in the future for stabilization of this language feature. The intention here
+is to make sure everyone's on board with the *general idea* of
+coroutines/generators being added to the Rust compiler and available for use on
+the nightly channel.
+
+# Motivation
+[motivation]: #motivation
+
+One of Rust's 2017 roadmap goals is ["Rust should be well-equipped for writing
+robust, high-scale servers"][goal]. A [recent survey][survey] has shown that
+the biggest blocker to robust, high-scale servers is ergonomic usage of async
```

Comment:
> Async I/O avoids context switches, which usually mean a call into the kernel and shuffling registers to sleep, and then another one along the same lines to wake back up.

Yes, context switching is relatively expensive. But procedure calls have costs as well (in callee-saved registers and spilling in general), so going non-blocking has its own version of context switching. The major unique cost to kernel threads is in FPU state, at least on x86.

> You're also going to be thrashing the CPU cache by completely switching out the stack hundreds of thousands of times a second or more.

Swap "the stack" for "the async function's state machine" and a similar issue applies. What goes on the stack in a blocking function ends up in the state machine of a stackless coroutine. Neither of them is touched until I/O completes.

> There's an issue with combining tens of thousands of I/O threads and a few compute threads that's hard to get right: either your compute threads get high priority and slow down I/O, or your compute threads get low priority and get preempted like hell.

I agree, preemption is the major cost. But no single approach appears to meet all of those requirements! Under heavy enough I/O demand a stackless coroutine must be split across multiple threads (requiring a heavier scheduler resembling what kernels already do) to avoid being CPU-starved. Cooperative threading can starve I/O threads by doing long computations, and preemptive threading (OS threads included) will generally be fair with CPU utilization but incur additional costs in possibly-unwanted context switches. OS thread schedulers are designed to be fair (achieving maximal utilization of all resources) under the majority of workloads, and choosing to use a less sophisticated scheduler than that provided by your OS should be a decision made with care.

---

> This is either a historical quirk or because OS vendors think you can do better than they can with application-specific scheduling.

In some situations, yes. I agree it is possible to do better than OS thread scheduling for some subset of all applications.

The question to answer here is whether async/await as one of at least three possible I/O concurrency models (disregarding an explicit event loop model because it has worse ergonomics by far) should be blessed by the Rust compiler, observing that the other two can be (and have been) implemented purely as libraries. As before, the answer here may be "yes, because it requires compiler support to work," but I haven't looked into implementation considerations enough to answer that question.